File: 295-find-median-from-data-stream/295-find-median-from-data-stream.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

#[derive(Default)]
struct MedianFinder {
    left: BinaryHeap<i32>,
    right: BinaryHeap<Reverse<i32>>,
}

impl MedianFinder {

    fn new() -> Self {
        Default::default()
    }
    
    fn add_num(&mut self, num: i32) {
        let mut left = &mut self.left;
        let mut right = &mut self.right;
        
        left.push(num);
        right.push(Reverse(left.pop().unwrap()));
        
        if left.len() < right.len(){
            let Reverse(n) = right.pop().unwrap();
            left.push(n);
        }
    }
    
    fn find_median(&self) -> f64 {
        let left = &self.left;
        let right = &self.right;
        
        if left.len() == right.len(){
            let a = *left.peek().unwrap();
            let Reverse(b) = *right.peek().unwrap();
            (a + b) as f64 / 2_f64
        }else{
            *left.peek().unwrap() as f64
        }
    }
}

File: 295-find-median-from-data-stream/295-find-median-from-data-stream.py
import heapq

class MedianFinder:
    def __init__(self):
        self.min_heap = []
        self.max_heap = []
        self.len = 0

    def addNum(self, num: int) -> None:
        
        if self.len == 0:
            heapq.heappush(self.max_heap, -num)
            self.len+=1
            return 
        
        if num > -self.max_heap[0]:
            heapq.heappush(self.min_heap, num)
        else:
            heapq.heappush(self.max_heap, -num)
        self.len+=1
        
        # print(self.max_heap, self.min_heap)
        
        mid = floor(self.len / 2)
        
        if len(self.max_heap) > mid:
            num = heapq.heappop(self.max_heap) * -1
            heapq.heappush(self.min_heap, num)
            
        if len(self.min_heap) > len(self.max_heap):
            num = heapq.heappop(self.min_heap)
            heapq.heappush(self.max_heap, -num)
        
        
    def findMedian(self) -> float:
        
        if self.len % 2 == 0:
            return (-self.max_heap[0] + self.min_heap[0])/2
        else:
            return -self.max_heap[0]
